### PR TITLE
[clang-format] Add space after a word token

### LINKED ISSUE #92688: [clang-format] A space is missing after a word token
A space is missing after `xor` in the following case with the PR https://github.com/llvm/llvm-project/pull/90161.

Expected formatting:

```c
void test(void) {
  static void (*xor)(uint8_t *, size_t, uint8_t);
  xor = resolve_xor_x86();
}
```

Actual formatting:

```c
 void test(void) {
  static void (*xor)(uint8_t *, size_t, uint8_t);
  xor= resolve_xor_x86();
}
```

### PR DESCRIPTION
**Pull request by robincaloudis**
Closes https://github.com/llvm/llvm-project/issues/92688

### CODE REVIEW

**Comment by robincaloudis** [comment]
Even though this PR gives us the correct behavior, I do not think it is good. 

Since I am by no means an expert on Clang, a few questions arose
* Why is `xor` tokenized as unary operator even though it's a word token in C?
* How to properly distinguish between keyword difference in the C and C++ language?

**Comment by owenca** [comment]
> Since I am by no means an expert on Clang, a few questions arose
> 
> * As Clang formats the code correctly (no code changes necessary) when instead of `xor` a different function name e.g. `xooor` is used, I wonder why `xor` is tokenized as Unary operator in the first place in C?
> * Is C and C++ using the same tokenizations?

clang-format formats all C code as C++. Since `xor` is a C++ keyword, it's lexed as `tok::caret`, which is then erroneously annotated as `TT_UnaryOperator`. This bug was uncovered by #90161.

> * How to properly distinguish between the C and C++ language in Clang Format?

clang-format can't do it properly. @mydeveloperday, @HazardyKnusperkeks, and @rymiel may know more about why we didn't add `LK_C` to the [`Language`](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#language) option.

File: clang/lib/Format/TokenAnnotator.cpp [hunk 1]
```diff
@@ -5280,7 +5280,8 @@ bool TokenAnnotator::spaceRequiredBefore(const AnnotatedLine &Line,
     // handled.
     if (Left.is(tok::amp) && Right.is(tok::r_square))
       return Style.SpacesInSquareBrackets;
-    return Style.SpaceAfterLogicalNot && Left.is(tok::exclaim);
+    return (Style.SpaceAfterLogicalNot && Left.is(tok::exclaim)) ||
+           Right.is(TT_BinaryOperator);
```
Comment (owenca) [inline clang/lib/Format/TokenAnnotator.cpp:5284]: This line was deleted in #90161. Adding it back will hide the uncovered bug and only fix #92688 superficially IMO. For example, `SpaceBeforeAssignmentOperators` will have no effect.

**Comment by mydeveloperday** [comment]
Because if this code was in a .h and not a .c you wouldn't know what language you were in

**Comment by mydeveloperday** [comment]
I thought we had some code like this around using a variable called "try"

**Comment by HazardyKnusperkeks** [comment]
> > Since I am by no means an expert on Clang, a few questions arose
> > 
> > * As Clang formats the code correctly (no code changes necessary) when instead of `xor` a different function name e.g. `xooor` is used, I wonder why `xor` is tokenized as Unary operator in the first place in C?
> > * Is C and C++ using the same tokenizations?
> 
> clang-format formats all C code as C++. Since `xor` is a C++ keyword, it's lexed as `tok::caret`, which is then erroneously annotated as `TT_UnaryOperator`. This bug was uncovered by #90161.
> 
> > * How to properly distinguish between the C and C++ language in Clang Format?
> 
> clang-format can't do it properly. @mydeveloperday, @HazardyKnusperkeks, and @rymiel may know more about why we didn't add `LK_C` to the [`Language`](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#language) option.

If I remember correctly, I was in favor of adding a C language.


> Because if this code was in a .h and not a .c you wouldn't know what language you were in

There are certainly headers which are ambiguous, and we could add an option to set the language of such headers. But if we hit a `class`, `namespace`, or a `::` we could fairly certain use it as C++ header. Similar to `guessIsObjC`, and skip checking for C++ if the new option is set to C++ (which of course would be the default, to keep existing behavior).

**Comment by owenca** [comment]
See #92880.

**Comment by robincaloudis** [comment]
Thanks @owenca, @mydeveloperday and @HazardyKnusperkeks for the explanation and insights! I'm closing this issue as @owenca found a much better solution.

**Pull request closed without merging**